Course/Course/Presentation/Dates/CourseDatesViewModel.swift

```diff
@@ -140,3 +186,195 @@ extension CourseDatesViewModel {
+                    await self.addCourseEvents()
+                }
+            },
+            type: .addCalendar
+        )
+    }
+    
+    private func showRemoveCalendarAlert() {
+        router.presentAlert(
+            alertTitle: CourseLocalization.CourseDates.removeCalendarTitle,
+            alertMessage: CourseLocalization.CourseDates.removeCalendarPrompt(
+                config.platformName,
+                courseName
+            ),
+            positiveAction: CoreLocalization.Alert.accept,
+            onCloseTapped: {
+                self.router.dismiss(animated: true)
+            },
+            okTapped: {
+                self.router.dismiss(animated: true)
+                self.removeCourseCalendar { [weak self] success in
+                    guard let self else { return }
+                    calendarEventsRemoved = true
+                }
+                
+            },
+            type: .removeCalendar
+        )
+    }
+    
+    private func showEventsAddedSuccessAlert() {
+        if calendar.isModalPresented {
+            calendarEventsAdded = true
+            return
+        }
+        calendar.isModalPresented = true
+        router.presentAlert(
+            alertTitle: "",
+            alertMessage: CourseLocalization.CourseDates.datesAddedAlertMessage(
+                calendar.calendarName
+            ),
+            positiveAction: CourseLocalization.CourseDates.calendarViewEvents,
+            onCloseTapped: {
+                self.router.dismiss(animated: true)
+                self.isOn = true
+                self.calendar.syncOn = true
+            },
+            okTapped: {
+                self.router.dismiss(animated: true)
+                if let url = URL(string: "calshow://"), UIApplication.shared.canOpenURL(url) {
+                    UIApplication.shared.open(url, options: [:], completionHandler: nil)
+                }
+            },
+            type: .calendarAdded
+        )
+    }
+    
+    func showCalendarSyncProgressView() {
+        router.presentView(
+            transitionStyle: .crossDissolve,
+            view: CalendarSyncProgressView(title: CourseLocalization.CourseDates.calendarSyncMessage)
+        )
+    }
+    
+    @MainActor
+    private func addCourseEventsIfNecessary() {
+        Task {
+            if calendar.syncOn && calendar.checkIfEventsShouldBeShifted(for: courseDates?.dateBlocks ?? [:]) {
+                showCalendarEventShiftAlert()
+            }
+        }
+    }
+    
+    @MainActor
+    private func showCalendarEventShiftAlert() {
+        router.presentAlert(
+            alertTitle: CourseLocalization.CourseDates.calendarOutOfDate,
+            alertMessage: CourseLocalization.CourseDates.calendarShiftMessage,
+            positiveAction: CourseLocalization.CourseDates.calendarShiftPromptUpdateNow,
+            onCloseTapped: {
+                // Remove course calendar
+                self.router.dismiss(animated: true)
```

Comment:
How about using weak self in closures? if so please update it for all alert functions.